src/mon/PGMap.h

```diff
@@ -253,6 +253,17 @@ class PGMap {
     return pool_stat_t();
   }
 
+  int get_num_up_primary_pg_by_osd(int osd) const {
+    assert(osd >= 0);
+    int num = 0;
+    for (const auto &p : pg_stat) {
+      if (p.second.up_primary == osd)
+        ++num;
+    }
```

Comment:
On second thought, we should probably not calculate this each time.  For a large number of PGs this will be extremely slow.  Let's make a map<int,int> num_primary_pg_by_osd PGMap member to go along with pg_by_osd?  It can just be a count probably (not a full set<>).